### ISSUE TITLE
WishList: pre-build ycmd

### ISSUE THREAD

**Issue by boyw165** [abingham/emacs-ycmd#95]
Is it possible to have a pre-build ycmd (Ubuntu/FreeBSD/Mac OSX) in the repo?

**Comment by akirakyle** [comment]
Has any progress been made towards this? I just recently got company-ycmd up and running after abandoning it awhile back when it proved more difficult to install than other options. (However I found myself really missing the fuzzy completion and ranking of ycm). I just realized that irony-mode also uses a complied server component based on libclang, and as such is very similar to ycmd, however they manage to have a very simple install mechanism through MELPA, which basically asks you to run the irony-install-server lisp function upon first running irony. I thought you might be interested to look at it, and it would be awesome if something similar could be implemented for emacs-ycmd [https://github.com/Sarcasm/irony-mode/blob/master/irony.el#L457-L534](url)

**Comment by ptrv** [comment]
@akirakyle The thing is that in `irony-mode` the code for the server is part of the package and gets downloaded alongside the package when installing with `package.el` through `MELPA`. This makes it easier to to build it from within the package. The `ycmd` server is a different repo and needs to be checked out and we need more tools to build it (like git). I don't see much benefit over the effort in implementing and install command.

**Comment by abingham** [comment]
Like @ptrv, I think there's a lot of inherent difficulty in building the ycmd client on behalf of ycmd. We need to grab the code, find the right python version, ask about a number of build options (e.g. different completers), and manage all of this on several platforms. Right now it just feels easier to ask users to follow the ycmd instructions.

With that said, the [ycmd project page indicates that clients commonly build the server](https://github.com/Valloric/ycmd#building), so perhaps it's not as complex as I expect. Does anyone have experience with what other clients are doing?

**Comment by ptrv** [comment]
I checked the other [listed clients](https://github.com/Valloric/ycmd#known-ycmd-clients). It doesn't look like the clients build themselves. There is always a link to the YCM build instruction. Even with the original vim YCM plugin you have to build the server separately.

**Comment by abingham** [comment]
OK, thanks for checking that.

I say that we close this. If someone wants to put in the effort to create functions which can compile the ycmd server, then of course we should consider them. But that doesn't seem like a good focus for this project.

Perhaps ycmd will eventually have installers and so forth that will *really* simplify the installation process.

**Comment by ptrv** [comment]
> I say that we close this. If someone wants to put in the effort to create functions which can compile the ycmd server, then of course we should consider them. But that doesn't seem like a good focus for this project.

:+1:

**Comment by akirakyle** [comment]
Thanks for the clarification! This definitely makes sense, and I think part of my confusion when I first attempted to get `emacs-ycmd` up and running came from the fact that [the build section of the ycmd project page](https://github.com/Valloric/ycmd#building) does seem to imply that `emacs-ycmd` should have been building at setting up ycmd for me. As @ptrv mentioned, it doesn't look like any of the other clients they list build ycmd either, so I submitted a PR to remove that comment from the build section. Hopefully that will make things a little clearer for people in the future.
